backend/compress: add zstd compression

A1ex3 opened the pull request:
<!--
Thank you very much for contributing code or documentation to rclone! Please
fill out the following questions to make it easier for us to review your
changes.

You do not need to check all the boxes below all at once, feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box.
-->

#### What is the purpose of this change?

Added support for reading and writing seekable Zstd-compressed archives using [github.com/klauspost/compress/zstd](https://github.com/klauspost/compress) and [github.com/SaveTheRbtz/zstd-seekable-format-go/pkg](https://github.com/SaveTheRbtz/zstd-seekable-format-go) (ZSTD seekable compression format implementation in Go).

Updated `Go` version from 1.24.0 to 1.24.4 to meet the requirements of `github.com/SaveTheRbtz/zstd-seekable-format-go/pkg`.

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

A1ex3 committed:
feat(backend/compress): add seekable zstd compression

Added support for reading and writing zstd-compressed archives in seekable format
using "github.com/klauspost/compress/zstd" and
"github.com/SaveTheRbtz/zstd-seekable-format-go/pkg".

Bumped Go version from 1.24.0 to 1.24.4 due to requirements of
"github.com/SaveTheRbtz/zstd-seekable-format-go/pkg".

kapitainsky commented:
Great idea to introduce this modern compression algorithm. My question is why you limit compression levels to just 4 options? I think it would be much more useful to support all standard levels 1 to 19. I do not see any reason why to use only 4 arbitrary selected levels.

For example myself I often use level 17 for online backup software as its speed on my computer nicely matches my max Internet uplink speed. I do not see why to remove option for fine tuning like this when it is available.

A1ex3 commented:
The `Go` zstd library [github.com/klauspost/compress](https://github.com/klauspost/compress) supports only four compression levels:
```go
// EncoderLevel predefines encoder compression levels.
// Only use the constants made available, since the actual mapping
// of these values are very likely to change and your compression could change
// unpredictably when upgrading the library.
type EncoderLevel int

const (
	speedNotSet EncoderLevel = iota

	// SpeedFastest will choose the fastest reasonable compression.
	// This is roughly equivalent to the fastest Zstandard mode.
	SpeedFastest

	// SpeedDefault is the default "pretty fast" compression option.
	// This is roughly equivalent to the default Zstandard mode (level 3).
	SpeedDefault

	// SpeedBetterCompression will yield better compression than the default.
	// Currently it is about zstd level 7-8 with ~ 2x-3x the default CPU usage.
	// By using this, notice that CPU usage may go up in the future.
	SpeedBetterCompression

	// SpeedBestCompression will choose the best available compression option.
	// This will offer the best compression no matter the CPU cost.
	SpeedBestCompression

	// speedLast should be kept as the last actual compression option.
	// The is not for external usage, but is used to keep track of the valid options.
	speedLast
)
```

The `EncoderLevelFromZstd` function allows converting a standard zstd level to one of the levels supported by the library. However, there is practically no benefit from this conversion:
```go
// EncoderLevelFromZstd will return an encoder level that closest matches the compression
// ratio of a specific zstd compression level.
// Many input values will provide the same compression level.
func EncoderLevelFromZstd(level int) EncoderLevel {
	switch {
	case level < 3:
		return SpeedFastest
	case level >= 3 && level < 6:
		return SpeedDefault
	case level >= 6 && level < 10:
		return SpeedBetterCompression
	default:
		return SpeedBestCompression
	}
}
```

A1ex3 committed:
fix(backend/compress): resolve golangci-lint warnings in backend/compress/szstd_helper.go

A1ex3 committed:
fix(backend/compress): handle "Uncompressed" case in "isCompressible" function

ncw reviewed (commented):
This looks like a useful feature - thank you :-)

All those switch statements

```go
switch {
	case Gzip:
	case Zstd:
	default:
}
```

Have made the code rather untidy though.

What do you think about factoring it into two implementations, defined by an interface. You can then separate the implementations into zstd.go and gzip.go (say)

That should restore a bit of sanity to the code.

A1ex3 committed:
feat(backend/compress): refactor compression handling via interface

- Replaced switch-based handling with interface-driven design
- Added separate implementations: gzip_handler.go, zstd_handler.go, uncompressed_handler.go and unknown_handler.go

A1ex3 committed:
feat(backend/compress/zstd): add "TestCompressZstd" target, refactor import, fix metadata, update deps

- Add "TestCompressZstd" target in "fstest/test_all/config.yaml"
- Refactor metadata handling
- Fix metadata size reference in Zstd compression handler
- Update package import from "github.com/SaveTheRbtz/zstd-seekable-format-go/pkg@v0.8.0" to "github.com/a1ex3/zstd-seekable-format-go/pkg@v0.10.0"
- Lower minimal Go version from 1.24.4 to 1.24.0

A1ex3 commented:
Hi,
I’ve just merged the latest changes from master locally into `feat-compress-zstd` branch and re-ran the integration tests for `TestCompress` and `TestCompressZstd` - both passed successfully.
Could you please share the test report or steps to reproduce the failing bisync tests?

roucc reviewed (approved):
Hi @A1ex3, I've rebased this and ran the tests and they all pass. Not sure what the problem with the bisync tests was before but @ncw did some work to fix them yesterday. Sorry this has taken so long and I'll merge this once the CI passes, thanks!

kapitainsky commented:
@A1ex3 could you please have a look at:

https://forum.rclone.org/t/compress-findings/53859

and provide your feedback?

gordan-bobic commented:
This is expected. Here are some related testing results from ZFS back when they implemented zstd. You need at least zstd-6 to match gzip-9, but zstd is massively faster.
<img width="2109" height="1113" alt="zfs-compression-2" src="https://github.com/user-attachments/assets/a49a5a4b-63c0-4c69-b0ae-601e75613648" />

Regarding what the compression levels mean, I'm going to hazard a guess they are directly 1:1 mappings onto what the underlying zstd golang library does:
https://pkg.go.dev/github.com/klauspost/compress/zstd
So your question should probably be directed here:
https://github.com/klauspost/compress

kapitainsky commented:
> This is expected.

Sorry but what is expected? I have tested using stand-alone gzip and zstd and this golang implementation behaves at least weird. (see details in the linked forum thread)

I have used silesia files set for test and already zstd -3 produces better results than gzip -9. But in case of this golang implementation even `SpeedBestCompression` is worse that gzip. Does it mean that it is not using even zstd level 3?

gordan-bobic commented:
Like I said - your questions on this should really be directed here:
https://github.com/klauspost/compress
